File: typescript/packages/v2-preview/src/batch.ts
/**
 * v2 Pillar 2 — Batch evaluation types.
 *
 * Structural mirrors of `contract/schemas/v2/evaluate-batch-request.schema.json`
 * and `contract/schemas/v2/evaluate-batch-response.schema.json`. Field
 * names stay snake_case (wire-identical) for the same reason
 * Pillar 9's `Proof` does — byte-level tooling shouldn't need a
 * translation layer. SDK-side camelCase wrappers layer on top in
 * the eventual v2 client.
 *
 * Per-item response is a discriminated union on `permitted`:
 * `permitted === true` narrows to the allow shape (with non-empty
 * `decision_id`); `permitted === false` keeps the same fields with
 * `reason` typically set to a denial explanation.
 */

/** Maximum items in a single batch — matches `evaluate-batch-request.schema.json`. */
export const EVALUATE_BATCH_MAX_ITEMS = 1000;

/** Pillar 9 proof status reported on a batch item that opted in via `payload_hash`. */
export type BatchProofStatus =
  | "pending"
  | "executed"
  | "failed"
  | "not_applicable";

/** One inbound evaluate request inside a batch. */
export interface BatchEvaluateItem {
  /** Same semantics as `/v1-evaluate-request.action`. */
  action: string;
  /** Same semantics as `/v1-evaluate-request.agent`. */
  agent: string;
  /** Same semantics as `/v1-evaluate-request.context`. */
  context: Record<string, unknown>;
  /**
   * Optional client-side payload hash. When set, the corresponding
   * response item carries `proof_id` + `proof_status` so a later
   * `consume` call can close the Pillar 9 proof loop.
   */
  payload_hash?: string;
  /** Optional target resource identifier. Mirrors `Proof.target`. */
  target?: string;
}

/** Request body for `POST /v2/evaluate:batch`. */
export interface EvaluateBatchRequest {
  requests: BatchEvaluateItem[];
  /** Echoed for wire parity with `/v1-evaluate`. Auth is via Bearer header. */
  api_key: string;
}

/**
 * Common fields every per-item response carries. Discriminator is
 * `permitted`; allow / deny narrow off it.
 */
interface BatchEvaluateResponseItemCommon {
  /** Zero-based position in the inbound `requests` array. */
  index: number;
  /** Per-item permit id. Same semantics as v1 `EvaluateResponse.decision_id`. */
  decision_id: string;
  /** Human-readable explanation. Empty string if the engine has nothing to add. */
  reason: string;
  /** Per-item audit-chain hash. Present even on deny. */
  audit_hash: string;
  /** ISO 8601 timestamp of this item's decision. */
  timestamp: string;
  /** Echo of the top-level batch_id, repeated per item for parallel post-processing. */
  batch_id: string;
  /** Pillar 9 proof id — present iff the request item carried `payload_hash`. */
  proof_id?: string;
  /** Pillar 9 proof status — present iff the request item carried `payload_hash`. */
  proof_status?: BatchProofStatus;
}

/** Per-item response when `permitted === true`. */
export interface BatchEvaluateAllowItem extends BatchEvaluateResponseItemCommon {
  permitted: true;
}

/** Per-item response when `permitted === false`. */
export interface BatchEvaluateDenyItem extends BatchEvaluateResponseItemCommon {
  permitted: false;
}

/** Discriminated union over allow / deny per-item responses. */
export type BatchEvaluateResponseItem =
  | BatchEvaluateAllowItem
  | BatchEvaluateDenyItem;

/** Response body for `POST /v2/evaluate:batch`. */
export interface EvaluateBatchResponse {
  batch_id: string;
  items: BatchEvaluateResponseItem[];
}

File: python/atlasent_v2_preview/atlasent_v2_preview/batch.py
"""v2 Pillar 2 — Batch evaluation models + builders.

Python sibling of ``typescript/packages/v2-preview/src/batch.ts``
plus ``buildBatch.ts``. Pure data — no HTTP. The v2-preview package
stays HTTP-free; these helpers let callers wire their own httpx /
requests client with consistent shape checking.

Per-item response is a discriminated union on ``permitted``:
``BatchEvaluateAllowItem`` carries the allow shape;
``BatchEvaluateDenyItem`` is the deny shape. ``isinstance()``
narrows in the same way TypeScript narrows on the ``permitted``
literal.
"""

from __future__ import annotations

import re
from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

EVALUATE_BATCH_MAX_ITEMS = 1000
"""Maximum items in a single batch — matches the schema's ``maxItems``."""


BatchProofStatus = Literal["pending", "executed", "failed", "not_applicable"]
"""Pillar 9 proof status reported on a batch item that opted in via ``payload_hash``."""


_HASH_HEX = re.compile(r"^[0-9a-f]{64}$")


class BatchEvaluateItem(BaseModel):
    """One inbound evaluate request inside a batch.

    See ``contract/schemas/v2/evaluate-batch-request.schema.json``
    ``$defs.BatchEvaluateItem``.
    """

    action: str = Field(min_length=1, max_length=256)
    agent: str = Field(min_length=1, max_length=256)
    context: dict[str, Any] = Field(default_factory=dict)
    payload_hash: str | None = Field(default=None)
    """Optional client-side payload hash. Opt-in to the Pillar 9 proof flow."""

    target: str | None = None
    """Optional target resource identifier. Mirrors ``Proof.target``."""

    model_config = {"extra": "allow"}

    @model_validator(mode="after")
    def _validate_payload_hash(self) -> BatchEvaluateItem:
        if self.payload_hash is not None and not _HASH_HEX.match(self.payload_hash):
            raise ValueError(
                "payload_hash must be 64 lowercase hex chars"
            )
        return self


class EvaluateBatchRequest(BaseModel):
    """Request body for ``POST /v2/evaluate:batch``.

    See ``contract/schemas/v2/evaluate-batch-request.schema.json``.
    """

    requests: list[BatchEvaluateItem] = Field(
        min_length=1, max_length=EVALUATE_BATCH_MAX_ITEMS
    )
    api_key: str = Field(min_length=1)

    model_config = {"extra": "forbid"}


class _ResponseItemBase(BaseModel):
    """Common fields every per-item response carries.

    ``permitted`` discriminates allow vs. deny.
    """

    index: int = Field(ge=0)
    decision_id: str
    reason: str = ""
    audit_hash: str
    timestamp: str
    batch_id: str
    proof_id: str | None = None
    proof_status: BatchProofStatus | None = None

    model_config = {"extra": "allow"}


class BatchEvaluateAllowItem(_ResponseItemBase):
    """Per-item response when ``permitted is True``."""

    permitted: Literal[True]


class BatchEvaluateDenyItem(_ResponseItemBase):
    """Per-item response when ``permitted is False``."""

    permitted: Literal[False]


BatchEvaluateResponseItem = BatchEvaluateAllowItem | BatchEvaluateDenyItem
"""Discriminated union — narrow via ``isinstance()`` on either variant."""


class EvaluateBatchResponse(BaseModel):
    """Response body for ``POST /v2/evaluate:batch``.

    See ``contract/schemas/v2/evaluate-batch-response.schema.json``.
    """

    batch_id: str = Field(min_length=1)
    items: list[BatchEvaluateAllowItem | BatchEvaluateDenyItem] = Field(
        default_factory=list
    )

    model_config = {"extra": "allow"}


# ── Builders ─────────────────────────────────────────────────────────


def build_evaluate_batch_request(
    items: list[BatchEvaluateItem | dict[str, Any]],
    api_key: str,
) -> EvaluateBatchRequest:
    """Construct a wire-valid :class:`EvaluateBatchRequest`.

    Accepts either fully-validated :class:`BatchEvaluateItem` instances
    or raw dicts (validated on the way in). Raises :class:`ValueError`
    on size or shape violations — the SDK boundary should catch them
    before sending an obviously malformed batch.
    """
    if not items:
        raise ValueError(
            "build_evaluate_batch_request: requests must contain at least 1 item"
        )
    if len(items) > EVALUATE_BATCH_MAX_ITEMS:
        raise ValueError(
            f"build_evaluate_batch_request: requests exceeds max "
            f"{EVALUATE_BATCH_MAX_ITEMS} (got {len(items)})"
        )
    if not isinstance(api_key, str) or not api_key:
        raise ValueError(
            "build_evaluate_batch_request: api_key must be a non-empty string"
        )
    cleaned: list[BatchEvaluateItem] = []
    for i, item in enumerate(items):
        if isinstance(item, BatchEvaluateItem):
            cleaned.append(item)
            continue
        try:
            cleaned.append(BatchEvaluateItem.model_validate(item))
        except Exception as err:
            raise ValueError(
                f"build_evaluate_batch_request: items[{i}] failed validation: {err}"
            ) from err
    return EvaluateBatchRequest(requests=cleaned, api_key=api_key)


def parse_evaluate_batch_response(body: Any) -> EvaluateBatchResponse:
    """Validate + narrow a server response body.

    Lifts the per-item discriminated union into the type system so
    callers can ``isinstance(item, BatchEvaluateAllowItem)`` /
    ``BatchEvaluateDenyItem`` without unsafe casts. Raises
    :class:`ValueError` on structural violations.
    """
    if not isinstance(body, dict):
        raise ValueError(
            "parse_evaluate_batch_response: body must be a JSON object"
        )
    try:
        return EvaluateBatchResponse.model_validate(body)
    except Exception as err:
        raise ValueError(
            f"parse_evaluate_batch_response: invalid response body: {err}"
        ) from err
